From eeeaf0fe2df339889f30f166c0b3c488535700cb Mon Sep 17 00:00:00 2001
From: Florian Weimer <fweimer@redhat.com>
Date: Fri, 19 Apr 2024 14:38:17 +0200
Subject: [PATCH 171/175] login: Check default sizes of structs utmp, utmpx,
 lastlog

The default <utmp-size.h> is for ports with a 64-bit time_t.
Ports with a 32-bit time_t or with __WORDSIZE_TIME64_COMPAT32=1
need to override it.

Reviewed-by: Adhemerval Zanella  <adhemerval.zanella@linaro.org>
(cherry picked from commit 4d4da5aab936504b2d3eca3146e109630d9093c4)
---
 login/Makefile                 |  2 +-
 login/tst-utmp-size.c          | 33 +++++++++++++++++++++++++++++++++
 sysdeps/arc/utmp-size.h        |  3 +++
 sysdeps/arm/utmp-size.h        |  2 ++
 sysdeps/csky/utmp-size.h       |  2 ++
 sysdeps/generic/utmp-size.h    | 23 +++++++++++++++++++++++
 sysdeps/hppa/utmp-size.h       |  2 ++
 sysdeps/m68k/utmp-size.h       |  3 +++
 sysdeps/microblaze/utmp-size.h |  2 ++
 sysdeps/mips/utmp-size.h       |  2 ++
 sysdeps/nios2/utmp-size.h      |  2 ++
 sysdeps/or1k/utmp-size.h       |  3 +++
 sysdeps/powerpc/utmp-size.h    |  2 ++
 sysdeps/riscv/utmp-size.h      |  2 ++
 sysdeps/sh/utmp-size.h         |  2 ++
 sysdeps/sparc/utmp-size.h      |  2 ++
 sysdeps/x86/utmp-size.h        |  2 ++
 17 files changed, 88 insertions(+), 1 deletion(-)
 create mode 100644 login/tst-utmp-size.c
 create mode 100644 sysdeps/arc/utmp-size.h
 create mode 100644 sysdeps/arm/utmp-size.h
 create mode 100644 sysdeps/csky/utmp-size.h
 create mode 100644 sysdeps/generic/utmp-size.h
 create mode 100644 sysdeps/hppa/utmp-size.h
 create mode 100644 sysdeps/m68k/utmp-size.h
 create mode 100644 sysdeps/microblaze/utmp-size.h
 create mode 100644 sysdeps/mips/utmp-size.h
 create mode 100644 sysdeps/nios2/utmp-size.h
 create mode 100644 sysdeps/or1k/utmp-size.h
 create mode 100644 sysdeps/powerpc/utmp-size.h
 create mode 100644 sysdeps/riscv/utmp-size.h
 create mode 100644 sysdeps/sh/utmp-size.h
 create mode 100644 sysdeps/sparc/utmp-size.h
 create mode 100644 sysdeps/x86/utmp-size.h

diff --git a/login/Makefile b/login/Makefile
index 62440499bcaeaa17dae31b04bb25cac6dcc3c675..ffe9f381148be247ea0d87ef72010c74fca897f6 100644
--- a/login/Makefile
+++ b/login/Makefile
@@ -44,7 +44,7 @@ subdir-dirs = programs
 vpath %.c programs
 
 tests := tst-utmp tst-utmpx tst-grantpt tst-ptsname tst-getlogin tst-updwtmpx \
-  tst-pututxline-lockfail tst-pututxline-cache
+  tst-pututxline-lockfail tst-pututxline-cache tst-utmp-size
 
 # Empty compatibility library for old binaries.
 extra-libs      := libutil
diff --git a/login/tst-utmp-size.c b/login/tst-utmp-size.c
new file mode 100644
index 0000000000000000000000000000000000000000..1b7f7ff04224efb530895614ec12e9f01b7aaa38
--- /dev/null
+++ b/login/tst-utmp-size.c
@@ -0,0 +1,33 @@
+/* Check expected sizes of struct utmp, struct utmpx, struct lastlog.
+   Copyright (C) 2024 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include <utmp.h>
+#include <utmpx.h>
+#include <utmp-size.h>
+
+static int
+do_test (void)
+{
+  _Static_assert (sizeof (struct utmp) == UTMP_SIZE, "struct utmp size");
+  _Static_assert (sizeof (struct utmpx) == UTMP_SIZE, "struct utmpx size");
+  _Static_assert (sizeof (struct lastlog) == LASTLOG_SIZE,
+                  "struct lastlog size");
+  return 0;
+}
+
+#include <support/test-driver.c>
diff --git a/sysdeps/arc/utmp-size.h b/sysdeps/arc/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..a247fcd3dab15f81d8878a2595ef980c1131763b
--- /dev/null
+++ b/sysdeps/arc/utmp-size.h
@@ -0,0 +1,3 @@
+/* arc has less padding than other architectures with 64-bit time_t.  */
+#define UTMP_SIZE 392
+#define LASTLOG_SIZE 296
diff --git a/sysdeps/arm/utmp-size.h b/sysdeps/arm/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/arm/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/csky/utmp-size.h b/sysdeps/csky/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/csky/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/generic/utmp-size.h b/sysdeps/generic/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..89dbe878b02301e918bfad7eb344459ff824c162
--- /dev/null
+++ b/sysdeps/generic/utmp-size.h
@@ -0,0 +1,23 @@
+/* Expected sizes of utmp-related structures stored in files.  64-bit version.
+   Copyright (C) 2024 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+/* Expected size, in bytes, of struct utmp and struct utmpx.  */
+#define UTMP_SIZE 400
+
+/* Expected size, in bytes, of struct lastlog.  */
+#define LASTLOG_SIZE 296
diff --git a/sysdeps/hppa/utmp-size.h b/sysdeps/hppa/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/hppa/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/m68k/utmp-size.h b/sysdeps/m68k/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..5946685819d6028979190bd619d40f57285ae451
--- /dev/null
+++ b/sysdeps/m68k/utmp-size.h
@@ -0,0 +1,3 @@
+/* m68k has 2-byte alignment.  */
+#define UTMP_SIZE 382
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/microblaze/utmp-size.h b/sysdeps/microblaze/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/microblaze/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/mips/utmp-size.h b/sysdeps/mips/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/mips/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/nios2/utmp-size.h b/sysdeps/nios2/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/nios2/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/or1k/utmp-size.h b/sysdeps/or1k/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..6b3653aa4dccd59d85c482d86b5f16f98f922e0e
--- /dev/null
+++ b/sysdeps/or1k/utmp-size.h
@@ -0,0 +1,3 @@
+/* or1k has less padding than other architectures with 64-bit time_t.  */
+#define UTMP_SIZE 392
+#define LASTLOG_SIZE 296
diff --git a/sysdeps/powerpc/utmp-size.h b/sysdeps/powerpc/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/powerpc/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/riscv/utmp-size.h b/sysdeps/riscv/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/riscv/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/sh/utmp-size.h b/sysdeps/sh/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/sh/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/sparc/utmp-size.h b/sysdeps/sparc/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/sparc/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
diff --git a/sysdeps/x86/utmp-size.h b/sysdeps/x86/utmp-size.h
new file mode 100644
index 0000000000000000000000000000000000000000..8f21ebe1b6c26ea1a5005bc2d0997ebf3b1d9bdd
--- /dev/null
+++ b/sysdeps/x86/utmp-size.h
@@ -0,0 +1,2 @@
+#define UTMP_SIZE 384
+#define LASTLOG_SIZE 292
-- 
2.35.6

